### PR TITLE
Fix a number of reordering issues in shift decomposition.

### PR DESCRIPTION
**Pull request by pgavlin**
Shift decompostition contained a fair number of IR rewrites that were
unintentionally reordering nodes. This change fixes these instances s.t.
they no longer reorder code.

### CODE REVIEW

**Comment by pgavlin** [comment]
@BruceForstall @mikedn @dotnet/jit-contrib PTAL

File: src/jit/decomposelongs.cpp [hunk 1]
```diff
@@ -1048,15 +1058,27 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
...
                 if (count < 32)
                 {
-                    // Hi is a GT_LSH_HI, lo is a GT_LSH. Will produce:
-                    // reg1 = lo
-                    // shl lo, shift
-                    // shld hi, reg1, shift
+                    // For shifts of < 32 bits, we transform the code to:
+                    //
+                    //     tLo = ...
+                    //           st.lclVar vLo, tLo
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Nit: IMO using the actual `genTreeOps` member names (STORE_LCL_VAR) would be preferable.

File: src/jit/decomposelongs.cpp [hunk 2]
```diff
@@ -1013,15 +1013,25 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
 {
     assert(use.IsInitialized());
 
-    GenTree* tree      = use.Def();
-    GenTree* gtLong    = tree->gtGetOp1();
+    GenTree* shift     = use.Def();
```
Comment (mikedn) [inline src/jit/decomposelongs.cpp]: Hurray for using descriptive variable names! But this should have been done separately, a lot of changes in this diff are caused by this and not by the actual bug fix.

**Comment by mikedn** [comment]
SGTM, I'm not familiar with this code. It looks like the main issue was that the original source nodes were moved right before the shift which is both incorrect and unnecessary.

**Comment by pgavlin** [comment]
>  It looks like the main issue was that the original source nodes were moved right before the shift which is both incorrect and unnecessary.

Yes, that was the problem in most cases. Oddly enough, some cases were also moving nodes earlier in the function.

File: src/jit/decomposelongs.cpp (hunk 2 above)
Comment (pgavlin) [inline src/jit/decomposelongs.cpp]: > But this should have been done separately, a lot of changes in this diff are caused by this and not by the actual bug fix.

Yes, you're right. In this case I elected to fold the changes together b/c it helped me reason about the change, but I should have undone the renames before submitting the PR.

**Comment by pgavlin** [comment]
@dotnet-bot help

**Comment by pgavlin** [comment]
@dotnet-bot
test Windows_NT x86 corefx_baseline
test Windows_NT x86 corefx_jitstress1
test Windows_NT x86 corefx_jitstress2

File: src/jit/decomposelongs.cpp [hunk 3]
```diff
@@ -1013,15 +1013,25 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
...
+    // tLo = ...
+    // ...
+    // tHi = ...
+    // ...
+    // tLong = long tLo, tHi
+    // ...
+    // tShiftAmount = ...
+    // ...
+    // tShift = shift tLong, tShiftAmount
+
```
Comment (briansull) [inline src/jit/decomposelongs.cpp]: This comment is confusing to me. Why do you use names like tLo, tHi  and tLong here.
These names should match your chosen local variables.
Also the "..." could be filled in, But I'm not sure that we even need this comment here, 
unless it is explaining something better than the actual code itself.

**Pull request merged**